bindings/ruby/src/lib.rs

```diff
@@ -15,16 +15,113 @@
 // under the License.
 
-use magnus::{define_global_function, function, Error};
-use opendal::{services::Memory, Operator};
+use std::{collections::HashMap, str::FromStr};
+
+use magnus::{
+    class, define_class, define_global_function, error::Result, exception, function, method,
+    prelude::*, Error,
+};
+use opendal::services::Memory;
 
 fn hello_opendal() {
-    let op = Operator::new(Memory::default()).unwrap().finish();
+    let op = opendal::Operator::new(Memory::default()).unwrap().finish();
     println!("{op:?}")
 }
 
+fn build_operator(
+    scheme: opendal::Scheme,
+    map: HashMap<String, String>,
+) -> Result<opendal::Operator> {
+    use opendal::services::*;
+
+    let op = match scheme {
+        opendal::Scheme::Azblob => opendal::Operator::from_map::<Azblob>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Azdfs => opendal::Operator::from_map::<Azdfs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Fs => opendal::Operator::from_map::<Fs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Gcs => opendal::Operator::from_map::<Gcs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Ghac => opendal::Operator::from_map::<Ghac>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Http => opendal::Operator::from_map::<Http>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Ipmfs => opendal::Operator::from_map::<Ipmfs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Memory => opendal::Operator::from_map::<Memory>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Obs => opendal::Operator::from_map::<Obs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Oss => opendal::Operator::from_map::<Oss>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::S3 => opendal::Operator::from_map::<S3>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Webdav => opendal::Operator::from_map::<Webdav>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        opendal::Scheme::Webhdfs => opendal::Operator::from_map::<Webhdfs>(map)
+            .map_err(format_magnus_error)?
+            .finish(),
+        _ => {
+            return Err(format_magnus_error(opendal::Error::new(
+                opendal::ErrorKind::Unexpected,
+                "not supported scheme",
+            )))
+        }
+    };
+
+    Ok(op)
+}
+
+#[magnus::wrap(class = "Operator", free_immediately, size)]
+#[derive(Clone, Debug)]
+pub struct Operator(opendal::BlockingOperator);
```

Comment:
So we will have an `AsyncOperator`?

Reply:
Oh, I's asking about the naming style. How ruby community naming a start which have both blocking and async API?

Reply:
I don't think ruby have the builtin await/async style async support, however, we can make api like the callback style:

```
operator = AsyncOperator.new
operator.write("blah") do |resp|
  puts resp.status
end
```

Reply:
I used to see code like the following:

```ruby
require 'async/io'

def echo_server(endpoint)
  Async do |task|
    # This is a synchronous block within the current task:
    endpoint.accept do |client|
      # This is an asynchronous block within the current reactor:
      data = client.read

      # This produces out-of-order responses.
      task.sleep(rand * 0.01)

      client.write(data.reverse)
      client.close_write
    end
  end
end
```

Reply:
> I used to see code like the following:
> 
> ```ruby
> require 'async/io'
> 
> def echo_server(endpoint)
>   Async do |task|
>     # This is a synchronous block within the current task:
>     endpoint.accept do |client|
>       # This is an asynchronous block within the current reactor:
>       data = client.read
> 
>       # This produces out-of-order responses.
>       task.sleep(rand * 0.01)
> 
>       client.write(data.reverse)
>       client.close_write
>     end
>   end
> end
> ```

unfortunately this is not the standard library but a third party gem, ruby have many possibilities to make magic happen on how the code executes, hard to get how it works for me :(

Reply:
> in the most of times, I guess a blocking only API is ok in ruby.

Great, let's focus on blocking API first.